templates/fine-tune-llm_v2/cookbooks/optimize_cost/README.ipynb

```diff
@@ -0,0 +1,302 @@
+    "\n",
+    "**⏱️ Time to complete**: 30 minutes\n",
+    "\n",
+    "Making models train faster is directly related to reducing the cost. But that is not the only objective that you should optimize.\n",
+    "Some people may not have access to the latest and greatest Nvidia GPUs, but still want to fine-tune LLMs on cheaper GPUs like L4s or A10s. So within this limitation you also want to 1) unblock yourself (even if it means more $/flops) and 2) optimize for best configuration while maintaining the quality of the models.\n",
+    "\n",
+    "LLMForge is an internal library developed at Anyscale which unifies the fine-tuning experience across Anyscale products. This is the core framework that powers both the finetuning API and this template. In the template, the library enables you to also choose certain knobs to accelerate training of the LLM for your specific application. In this cookbook, we will discuss the parameters that are exposed that can help squeeze more flops out of GPUs and train models faster. "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Table of Content\n",
+    "\n",
+    "- [Overview](#overview)\n",
+    "- [How to benchmark things yourself?](#how-do-you-benchmark-yourself) \n",
+    "- [Configurations](#configurations)\n",
+    "    - [Instance type](#instance-type)\n",
+    "    - [Batch size and Context length](#batch-size-and-context-length)\n",
+    "    - [Deepspeed configs](#deepspeed-configs)\n",
+    "    - [Activation checkpointing](#activation-checkpointing)\n",
+    "    - [Gradient accumulation](#gradient-accumulation)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Overview\n",
+    "\n",
+    "Picking the correct type of GPUs and right combination of configuration is not easy to prescribe. It usually comes down to running benchmarks yourself and be able to quickly understand the trade-offs for speed and memory. There is a large collection of tutorials on the internets that can give you holistic view of how to think about performance of training LLMs, so that we do not too much into the details.\n",
+    "The following two are very good resources:\n",
+    "\n",
+    "- https://sumanthrh.com/post/distributed-and-efficient-finetuning/\n",
+    "- https://github.com/stas00/ml-engineering/blob/master/training/performance/README.md (more advanced)\n",
+    "\n",
+    "\n",
+    "When thinking about training models, there are two issues that come up in order of importance:\n",
+    "\n",
+    "- What parameters should I set to make sure I do not run out of GPU Memory (hit CUDA OOM) during training?\n",
+    "- What parameters would give me a better throughput and therefore reduce my cost?\n",
+    "\n",
+    "Now in order to fit a large model (say 70B sized model) the priority is first fitting the model. If the you cannot even train a model at batch size = 1, then trowing more hardware at it may not help. \n",
```

Comment:
```suggestion
    "Now in order to fit a large model (say 70B sized model) the priority is first fitting the model. If you cannot train the model even with batch size = 1, then throwing more hardware at it may not help. \n",
```
nit